File: src/SistemaQuiz.cpp
/*

Trabalho 01 da Disciplina de Análise e Projeto de Software

Autor: Miguel Miron Silva

*/


//Gerados pelo Gemini
#include <iostream>
#include <vector>
#include <string>
#include <memory>
//Adicionados pelo Kimi
#include <stack>
#include <chrono>
#include <stdexcept>

using namespace std;

// --- SINGLETON (Timer) ---
//Modernizado pelo Kimi
class ExamTimer {
private:
    chrono::steady_clock::time_point start; // chrono ainda precisa do prefixo ou de outro using
    chrono::seconds duration;
    ExamTimer() : start(chrono::steady_clock::now()), duration(3600) {}
public:
    static ExamTimer& getInstance() {
        static ExamTimer instance;
        return instance;
    }
    int getRemainingSeconds() {
        auto elapsed = chrono::steady_clock::now() - start;
        auto remaining = duration -
               chrono::duration_cast<chrono::seconds>(elapsed);
        return static_cast<int>(remaining.count()) > 0 ?
               static_cast<int>(remaining.count()) : 0;
    }
};

// --- STRATEGY (Correção) ---
class GradingStrategy {
public:
    virtual bool grade(string answer, string key) = 0;
    virtual ~GradingStrategy() = default;
};

class ExactMatchStrategy : public GradingStrategy {
public:
    bool grade(string answer, string key) override {
        return answer == key;
    }
};


// --- COMPOSITE (Estrutural) ---
//Modernizado pelo Kimi - Folha, Composto e Iterator adicionados
class ExamComponent {
public:
    virtual void display(int depth = 0) const = 0;
    virtual void add(shared_ptr<ExamComponent> c) {
        throw runtime_error("Incapaz de adicionar folha!"); 
    }
    virtual ~ExamComponent() = default;
};


// --- PRODUTOS ---
// Folha
class Question : public ExamComponent {
    string text;
    string key;
    shared_ptr<GradingStrategy> grader;
public:
    Question(const string& t,
             const string& k,
             shared_ptr<GradingStrategy> g)
        : text(t), key(k), grader(move(g)) {} // move ao invés de std::move

    void display(int depth = 0) const override {
        string indent(depth * 2, ' ');
        cout << indent << "Questao: " << text << "\n";
    }

    bool checkAnswer(const string& ans) const {
        return grader->grade(ans, key);
    }
};

// Composto
class ExamSection : public ExamComponent {
    vector<shared_ptr<ExamComponent>> children;
    string title;
public:
    explicit ExamSection(const string& t) : title(t) {}

    void add(shared_ptr<ExamComponent> c) override {
        children.push_back(move(c));
    }

    void display(int depth = 0) const override {
        string indent(depth * 2, ' ');
        cout << indent << "--- SECAO: " << title << " ---\n";
        for (const auto& c : children) c->display(depth + 1);
    }

    // Iterator real (oculta estrutura interna)
    class Iterator {
        using Iter = vector<shared_ptr<ExamComponent>>::iterator;
        Iter curr, end;
    public:
        Iterator(vector<shared_ptr<ExamComponent>>& vec)
            : curr(vec.begin()), end(vec.end()) {}
        bool hasNext() { return curr != end; }
        shared_ptr<ExamComponent> next() {
            if (!hasNext()) return nullptr;
            return *curr++;
        }
    };

    Iterator createIterator() {
        return Iterator(children);
    }
};

// --- FACTORY METHOD ---
class QuestionFactory {
public:
    static shared_ptr<Question> createMultipleChoice(string text, string key) {
        return make_shared<Question>(text + " (A/B/C/D)", key, make_shared<ExactMatchStrategy>());
    }
};

// --- MEMENTO (Estado) ---
//Modernizado pelo Kimi - adicionado respostas
class ExamMemento {
    int currentIndex;
    vector<string> answers;
    chrono::steady_clock::time_point ts;
public:
    ExamMemento(int idx,
                vector<string> ans)
        : currentIndex(idx),
          answers(move(ans)),
          ts(chrono::steady_clock::now()) {}

    int getIndex() const { return currentIndex; }
    const vector<string>& getAnswers() const {
        return answers;
    }
};


// ---SESSION ---
//Modernizado pelo Kimi - adicionado respostas
class ExamSession {
    int currentIndex = 0;
    vector<string> answers;
public:
    void answerQuestion(const string& ans) {
        if (currentIndex >= answers.size())
            answers.push_back(ans);
        else
            answers[currentIndex] = ans;
        ++currentIndex;
    }

    ExamMemento save() {
        return ExamMemento(currentIndex, answers);
    }

    void restore(const ExamMemento& m) {
        currentIndex = m.getIndex();
        answers = m.getAnswers();
    }

    int getCurrentIndex() const { return currentIndex; }
};


// --- BUILDER  ---
//Modificado pelo Gemini para gerar árvore
class ExamBuilder {
    shared_ptr<ExamSection> root;
    ExamComponent* currentScope; 
public:
    ExamBuilder(const string& title) {
        root = make_shared<ExamSection>(title);
        currentScope = root.get();
    }
    ExamBuilder& addSection(const string& name) {
        auto sec = make_shared<ExamSection>(name);
        currentScope->add(sec);
        return *this;
    }
    ExamBuilder& addQuestion(const string& text,
                             const string& key) {
        currentScope->add(QuestionFactory::createMultipleChoice(text, key));
        return *this;
    }
    shared_ptr<ExamSection> build() { return root; }
};


// --- CLIENTE ---
//Modernizado pelo Kimi
int main() {
   //Singleton
    auto& timer = ExamTimer::getInstance();
    cout << "Tempo restante: " << timer.getRemainingSeconds() << "s\n";

    //Builder + Composite
    ExamBuilder builder("Prova Final de C++");
    auto prova = builder
        .addSection("Logica")
        .addQuestion("Quanto e 2+2?", "4")
        .addQuestion("Quanto e 3*3?", "9")
        .addSection("Orientacao a Objetos")
        .addQuestion("O que e polimorfismo?", "Muitas formas")
        .build();

    //Display
    prova->display();

    //terator
    cout << "\nPercorrendo com Iterator:\n";
    auto it = dynamic_cast<ExamSection*>(prova.get())
                  ->createIterator();
    while (it.hasNext()) {
        auto comp = it.next();
        comp->display();
    }

    //Session + Memento
    ExamSession session;
    session.answerQuestion("4");
    session.answerQuestion("10"); // errado
    auto checkpoint = session.save();

    cout << "\n[Sistema caiu... Restaurando...]\n";
    session.restore(checkpoint);
    cout << "Restaurado para indice: " << session.getCurrentIndex() << "\n";
    
    return 0;
}

File: src/SistemaQuiz.java
import java.util.ArrayList;
import java.util.List;
import java.util.Iterator;
import java.time.Instant;
import java.time.Duration;

/*
Trabalho 01 da Disciplina de Análise e Projeto de Software

Autor: Miguel Miron Silva

Versão Java
*/

public class ExamSystem {

    // --- SINGLETON (Timer) ---
    static class ExamTimer {
        private static ExamTimer instance;
        private final Instant start;
        private final long durationSeconds;

        private ExamTimer() {
            this.start = Instant.now();
            this.durationSeconds = 3600; // 1 hora
        }

        // Lazy initialization (thread-safe simples)
        public static synchronized ExamTimer getInstance() {
            if (instance == null) {
                instance = new ExamTimer();
            }
            return instance;
        }

        public long getRemainingSeconds() {
            Instant now = Instant.now();
            long elapsed = Duration.between(start, now).getSeconds();
            long remaining = durationSeconds - elapsed;
            return Math.max(remaining, 0);
        }
    }

    // --- STRATEGY (Correção) ---
    interface GradingStrategy {
        boolean grade(String answer, String key);
    }

    static class ExactMatchStrategy implements GradingStrategy {
        @Override
        public boolean grade(String answer, String key) {
            return answer.equals(key);
        }
    }

    // --- COMPOSITE (Estrutural) ---
    static abstract class ExamComponent {
        public abstract void display(int depth);

        public void add(ExamComponent c) {
            throw new UnsupportedOperationException("Incapaz de adicionar em uma folha!");
        }
    }

    // --- PRODUTOS ---
    // Folha (Leaf)
    static class Question extends ExamComponent {
        private String text;
        private String key;
        private GradingStrategy grader;

        public Question(String text, String key, GradingStrategy grader) {
            this.text = text;
            this.key = key;
            this.grader = grader;
        }

        @Override
        public void display(int depth) {
            String indent = "  ".repeat(depth);
            System.out.println(indent + "Questao: " + text);
        }

        public boolean checkAnswer(String ans) {
            return grader.grade(ans, key);
        }
    }

    // Composto (Composite)
    static class ExamSection extends ExamComponent {
        private List<ExamComponent> children = new ArrayList<>();
        private String title;

        public ExamSection(String title) {
            this.title = title;
        }

        @Override
        public void add(ExamComponent c) {
            children.add(c);
        }

        @Override
        public void display(int depth) {
            String indent = "  ".repeat(depth);
            System.out.println(indent + "--- SECAO: " + title + " ---");
            for (ExamComponent c : children) {
                c.display(depth + 1);
            }
        }

        // Iterator customizado (expondo a lógica do padrão Iterator)
        public ComponentIterator createIterator() {
            return new ComponentIterator(children);
        }

        // Inner class para o Iterator
        static class ComponentIterator {
            private Iterator<ExamComponent> iterator;

            public ComponentIterator(List<ExamComponent> list) {
                this.iterator = list.iterator();
            }

            public boolean hasNext() {
                return iterator.hasNext();
            }

            public ExamComponent next() {
                return iterator.next();
            }
        }
    }

    // --- FACTORY METHOD ---
    static class QuestionFactory {
        public static Question createMultipleChoice(String text, String key) {
            // Em Java, passamos a instância da estratégia
            return new Question(text + " (A/B/C/D)", key, new ExactMatchStrategy());
        }
    }

    // --- MEMENTO (Estado) ---
    static class ExamMemento {
        private final int currentIndex;
        private final List<String> answers;
        private final Instant timestamp;

        public ExamMemento(int idx, List<String> ans) {
            this.currentIndex = idx;
            // Importante: Deep copy da lista para preservar o estado naquele momento
            this.answers = new ArrayList<>(ans);
            this.timestamp = Instant.now();
        }

        public int getIndex() {
            return currentIndex;
        }

        public List<String> getAnswers() {
            return answers;
        }
    }

    // --- SESSION (Originator) ---
    static class ExamSession {
        private int currentIndex = 0;
        private List<String> answers = new ArrayList<>();

        public void answerQuestion(String ans) {
            if (currentIndex >= answers.size()) {
                answers.add(ans);
            } else {
                answers.set(currentIndex, ans);
            }
            currentIndex++;
        }

        public ExamMemento save() {
            return new ExamMemento(currentIndex, answers);
        }

        public void restore(ExamMemento m) {
            this.currentIndex = m.getIndex();
            this.answers = new ArrayList<>(m.getAnswers()); // Restaura copiando
        }

        public int getCurrentIndex() {
            return currentIndex;
        }
    }

    // --- BUILDER ---
    static class ExamBuilder {
        private ExamSection root;
        private ExamComponent currentScope;

        public ExamBuilder(String title) {
            root = new ExamSection(title);
            currentScope = root;
        }

        public ExamBuilder addSection(String name) {
            ExamSection sec = new ExamSection(name);
            currentScope.add(sec);
            // Nota: No código original C++, o scope não mudava, então as seções ficavam 
            // como irmãs na raiz. Mantive esse comportamento.
            return this;
        }

        public ExamBuilder addQuestion(String text, String key) {
            currentScope.add(QuestionFactory.createMultipleChoice(text, key));
            return this;
        }

        public ExamSection build() {
            return root;
        }
    }

    // --- CLIENTE (Main) ---
    public static void main(String[] args) {
        // Singleton
        ExamTimer timer = ExamTimer.getInstance();
        System.out.println("Tempo restante: " + timer.getRemainingSeconds() + "s");

        // Builder + Composite
        ExamBuilder builder = new ExamBuilder("Prova Final de Java");
        ExamSection prova = builder
                .addSection("Logica")
                .addQuestion("Quanto e 2+2?", "4")
                .addQuestion("Quanto e 3*3?", "9")
                .addSection("Orientacao a Objetos")
                .addQuestion("O que e polimorfismo?", "Muitas formas")
                .build();

        // Display (Composite)
        prova.display(0);

        

        // Iterator
        System.out.println("\nPercorrendo com Iterator:");
        // Em Java não precisamos do dynamic_cast, pois sabemos que 'prova' é ExamSection
        ExamSection.ComponentIterator it = prova.createIterator();
        while (it.hasNext()) {
            ExamComponent comp = it.next();
            comp.display(0);
        }

        // Session + Memento
        ExamSession session = new ExamSession();
        session.answerQuestion("4");
        session.answerQuestion("10"); // resposta errada proposital
        
        System.out.println("\n[Salvando estado...]");
        ExamMemento checkpoint = session.save();

        System.out.println("[Sistema caiu... Restaurando...]");
        session.restore(checkpoint);
        System.out.println("Restaurado para indice: " + session.getCurrentIndex());
    }
}
